tests/asgi/test_asgi_servers.py

```diff
@@ -93,6 +96,246 @@ def test_sse(self, server_base_url):
         assert not events[-1]
 
 
+class TestWebSocket:
+
+    @pytest.mark.parametrize('explicit_close', [True, False])
+    @pytest.mark.parametrize('close_code', [None, 4321])
+    def test_hello(self, explicit_close, close_code, server_url_events_ws):
+        echo_expected = 'Check 1 - \U0001f600'
+
+        extra_headers = {'X-Command': 'recv'}
+
+        if explicit_close:
+            extra_headers['X-Close'] = 'True'
+
+        if close_code:
+            extra_headers['X-Close-Code'] = str(close_code)
+
+        async def do_test():
+            async with websockets.connect(
+                server_url_events_ws,
+                extra_headers=extra_headers,
+            ) as ws:
+                got_message = False
+
+                while True:
+                    try:
+                        # TODO: Why is this failing to decode on the other side? (raises an error)
+                        # TODO: Why does this cause Daphne to hang?
+                        await ws.send(f'{{"command": "echo", "echo": "{echo_expected}"}}')
+
+                        message_text = await ws.recv()
+                        message_echo = await ws.recv()
+                        message_binary = await ws.recv()
+                    except websockets.exceptions.ConnectionClosed as ex:
+                        if explicit_close and close_code:
+                            assert ex.code == close_code
+                        else:
+                            assert ex.code == 1000
+
+                        break
+
+                    got_message = True
+                    assert message_text == 'hello world'
+                    assert message_echo == echo_expected
+                    assert message_binary == b'hello\x00world'
+
+                assert got_message
+
+        testing.invoke_coroutine_sync(do_test)
+
+    @pytest.mark.parametrize('explicit_close', [True, False])
+    @pytest.mark.parametrize('close_code', [None, 4040])
+    def test_rejected(self, explicit_close, close_code, server_url_events_ws):
+        extra_headers = {'X-Accept': 'reject'}
+        if explicit_close:
+            extra_headers['X-Close'] = 'True'
+
+        if close_code:
+            extra_headers['X-Close-Code'] = str(close_code)
+
+        async def do_test():
+            try:
+                async with websockets.connect(server_url_events_ws, extra_headers=extra_headers):
+                    pass
+
+                pytest.fail('no error raised')
+
+            except websockets.exceptions.InvalidStatusCode as ex:
+                assert ex.status_code == 403
+
+        testing.invoke_coroutine_sync(do_test)
+
+    def test_missing_responder(self, server_url_events_ws):
+        server_url_events_ws += '/404'
+
+        async def do_test():
+            try:
+                async with websockets.connect(server_url_events_ws):
+                    pass
+
+                pytest.fail('no error raised')
```

Comment:
Could it be worth adding a `NOTE` why we aren't just using [`pytest.raises`](https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions)?

Reply:
👍  Refactored all but one instance of this pattern, and added a note to the remaining instance.